### PR TITLE
MAUI-3916 Custom filtering feature included in UG documentation

### PR DESCRIPTION
**Pull request by jeya-kasipandi**
Custom filtering support for autocomplete included in ug documentaton.

### CODE REVIEW

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 1]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+/// </summary>
+public class CityFilteringBehavior : IAutocompleteFilterBehavior
+{
+
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+**Step 2:** Then, implement the `GetMatchingIndexes` method of `IAutocompleteFilterBehavior` interface to create your own suggestion list (containing the indices of the filtered items) based on the text entered in the `Autocomplete` control that needs to be shown in the drop-down. The `GetMatchingIndexes` method contains following arguments:
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:258]: **Step 2:** Then, implement the `GetMatchingIndexes` method of the `IAutocompleteFilterBehavior` interface to create your suggestion list (containing the indices of the filtered items) based on the text entered in the `Autocomplete` control that needs to be shown in the drop-down. The `GetMatchingIndexes` method contains the following arguments.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 2]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+* `source` - The owner of the filter behavior, which holds information about `ItemsSource`  property and so on.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:260]: * `Source` - The owner of the filter behavior, which holds information about the `ItemsSource`  property and more.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 3]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+* `filterInfo` - Contains details about the text entered in `Autocomplete` control. Using this text, you can prepare suggestion list, which gets displayed in the drop-down list.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:261]: * `FilterInfo` - Contains details about the text entered in the `Autocomplete` control. Using this text, prepare a suggestion list, which gets displayed in the drop-down list.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 4]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+                             select item);
+
+        return await Task.FromResult(filteredItems);
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+**Step 3:** Applying custom filtering to the `Autocomplete` control by using the `FilterBehavior` property.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:285]: **Step 3:** Apply the custom filtering to the `Autocomplete` control by using the `FilterBehavior` property.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 5]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+{% endhighlight %}
+{% endtabs %}
+
+The following image demonstrates how to display cities in the drop-down based on the country name entered in the `Autocomplete` control.
+
+![.NET MAUI AutoComplete filter the items based on custom filtering logic.](Images/Searching_Filtering/Custom_filtering.png)
+
+### Choose default item to select
+
+When searching, the first item in the drop-down will be highlighted by default. Using the `SearchBehavior` property, you can customize the default highlighting behavior by using your custom selection logic to select the items based on your search criteria. The default value of `SearchBehavior` is `null`.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:310]: When searching, the first item in the drop-down will be highlighted by default. Using the `SearchBehavior` property, customize the default highlighting behavior by using your custom selection logic to select the items based on your search criteria. The default value of the `SearchBehavior` is `null.`

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 6]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+When searching, the first item in the drop-down will be highlighted by default. Using the `SearchBehavior` property, you can customize the default highlighting behavior by using your custom selection logic to select the items based on your search criteria. The default value of `SearchBehavior` is `null`. 
+
+**Step 1:** Create a class that derives from the `IAutocompleteSearchBehavior` interface.
+
+{% tabs %}
+{% highlight C# %}
+
+/// <summary>
+/// Represents a custom searching behavior for `Autocomplete` control.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:318]: /// Represents a custom searching behavior for the `Autocomplete` control.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 7]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+public class CapitalCitySearchingBehavior : Syncfusion.Maui.Inputs.IAutocompleteSearchBehavior
+
+{
+
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+**Step 2:** Then, implement the `GetHighlightIndex` method of the `IAutocompleteSearchBehavior` interface to initially select any item from the filtered list in the `Autocomplete` control drop-down. The `GetHighlightIndex` method contains the following arguments.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:330]: **Step 2:** Then, implement the `GetHighlightIndex` method of the `IAutocompleteSearchBehavior` interface to select any item from the filtered list in the `Autocomplete` control drop-down. The `GetHighlightIndex` method contains the following arguments.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 8]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+* `source` - The owner of the search behavior, which holds information about `ItemsSource` and so on.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:332]: * `Source` - The owner of the search behavior, which holds information about the `ItemsSource` and more.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 9]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+* `searchInfo` - Contains information about the filtered items based on the text entered in `Autocomplete` control. You can select item from the drop-down list using this list.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:333]: * `SearchInfo` - Contains information about the filtered items based on the text entered in the `Autocomplete` control. Select an item from the drop-down list using this list.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 10]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+{% endhighlight %}
+{% endtabs %}
+
+**Step 2:** Then, implement the `GetHighlightIndex` method of the `IAutocompleteSearchBehavior` interface to initially select any item from the filtered list in the `Autocomplete` control drop-down. The `GetHighlightIndex` method contains the following arguments.
+
+* `source` - The owner of the search behavior, which holds information about `ItemsSource` and so on.
+* `searchInfo` - Contains information about the filtered items based on the text entered in `Autocomplete` control. You can select item from the drop-down list using this list. 
+
+The following code demonstrates initially how to select an item from the filtered list of drop-down in the `Autocomplete` control. When entering the country's name, the capital cities will be selected.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:335]: The following code initially demonstrates how to select an item from the filtered list of the drop-down in the `Autocomplete` control. When entering the country's name, the capital cities will be selected.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 11]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+        return 0;
+    }
+}
+
+{% endhighlight %}
+
+{% endtabs %}
+
+**Step 3:** Applying custom searching to `Autocomplete` control by using the `SearchBehavior` property.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:359]: **Step 3:** Apply the custom searching to the `Autocomplete` control by using the `SearchBehavior` property.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 12]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+The following gif demonstrates how to select the capital city from the drop-down based on the country name entered in the `Autocomplete` control.
+
+![.NET MAUI AutoComplete filter the items based on custom filtering and searching logic.](Images/Searching_Filtering/Custom_filtering_searching.gif)
+
+## Load asynchronous items
+
+Load the data dynamically at runtime based on typed input. This dynamic loading can be done while performing custom filtering using the `CustomFilter` property.
+
+The `GetMatchingItemsAsync` method of `IAutocompleteFilterBehavior` helps you to perform filtering operations on different threads without blocking the current thread by using `await Task.Run()`.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:388]: The `GetMatchingItemsAsync` method of the `IAutocompleteFilterBehavior` helps you perform filtering operations on different threads without blocking the current thread by using `await Task.Run().`

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 13]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+**Step 1:** Create a class that derives from the `IAutocompleteFilterBehavior` interface and add your custom filter logic in `GetMatchingItemsAsync` method to load the run time items based on typed input.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:390]: **Step 1:** Create a class from the `IAutocompleteFilterBehavior` interface and add your custom filter logic to the `GetMatchingItemsAsync` method to load the runtime items based on typed input.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 14]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+
+            return list;
+        }, token);
+    }
+}
+
+{% endhighlight %}
+{% endtabs %}
+
+**Step 2:** Applying `CustomAsyncFilter` to `Autocomplete` control by using the `FilterBehavior` property.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:429]: **Step 2:** Apply the `CustomAsyncFilter` to the `Autocomplete` control by using the `FilterBehavior` property.

File: MAUI/Autocomplete/Searching-Filtering.md [hunk 15]
```diff
@@ -171,6 +171,276 @@ autocomplete.TextSearchMode = AutocompleteTextSearchMode.Contains;
...
+            </editors:SfAutocomplete.FilterBehavior>
+<editors:SfAutocomplete.SearchBehavior>
+                <local:CapitalCitySearchingBehavior></local:CapitalCitySearchingBehavior>
+            </editors:SfAutocomplete.SearchBehavior>
+</editors:SfAutoComplete>
+
+{% endhighlight %}
+{% endtabs %}
+
+The following gif demonstrates how to select the capital city from the drop-down based on the country name entered in the `Autocomplete` control.
```
Comment (Chrispineagunja) [inline MAUI/Autocomplete/Searching-Filtering.md:380]: The following image demonstrates how to select the capital city from the drop-down based on the country name entered in the `Autocomplete` control.

**Review by PaulAndersonS — APPROVED** [review]

**Pull request merged**